src/StateMachine.cpp

```diff
@@ -152,6 +163,11 @@ StateMachine::Update StateMachine::processSDOReceive(canbus::Message const& msg)
         if (msg.time.isNull()) {
             throw ProtocolError("received CAN message with zero timestamp");
         }
+        if (cmd.size == 0 && useUnknownSizes) {
+            if (!has(objectId, subId)) {
+                declareInternal(objectId, subId, 4)->second.knownSize = false;
+            }
+        }
         setObjectValue(objectId, subId, msg.time, msg.data + 4, cmd.size);
```

Comment:
I think it's important that the last argument to setObjectValue is indeed the number of valid bytes that can (unknown size) or should (valid size) be read from the incoming buffer. This makes sure that we can safely not do a buffer overflow while reading.

A.k.a. normalize cmd.size to 4 in the unknown size case.